modules/core/site.js

```diff
@@ -257,8 +257,12 @@ var Hm_Ajax_Request = function() { return {
                 Hm_Ajax.err_condition = false;
                 Hm_Notices.hide(true);
             }
-            if (res.router_user_msgs && !$.isEmptyObject(res.router_user_msgs)) {
-                Hm_Notices.show(res.router_user_msgs);
+            if (res.router_user_msgs && !$.isEmptyObject(res.router_user_msgs)) {             
+                var hm_msg = Hm_Utils.json_encode(res.router_user_msgs);
+                if (! Hm_Notices.displayed_messages.includes(hm_msg)) {
```

Comment:
I am not sure this is the proper fix. I suspect that multiple jquery events are stacking and executing together in order to receive several identical messages. This fix will hide the duplicates but not deal with other issues like submitting several requests to the server for one and the same thing - while delete might not cause issues, other repeatable operations like send might cause. Thus, we should check why do we receive several identical messages and fix the corresponding code - if we add the same event handlers without the page being reloaded we should first remove the old one before adding or don't add the handler at all.

Reply:
By the way, I found that messages are saved in a table in session. What JavaScript does is retrieve messages in session and display them.

 A solution came to mind, and if for each message we added an identifier, this could allow us to delete it in the message session table. What do you think ?

Reply:
This doesn't explain why do you see several "Message deleted" messages. Again, we don't want to make the messages disappear by checking their ID or similarity. We want to relay the messages coming from the server to the client but we need to make sure that no duplicate ajax calls are executed, even when multiple tabs are open. Please specify here an exact use-case that reproduced the problem and I can check if you can't find the reason for the duplicate messages.

Reply:
Here's how I figured out the bug:

Sending the request to delete the message: **_ajax_imap_delete_message_**(1)
    The time the request is being processed;
    The application sends other requests including:
    - **_ajax_imap_unsnooze_**(2)
    - **_ajax_imap_folder_display_**(3)
    - **_ajax_imap_unread_**(4)
    
Queries **_(2),(3), (4)_** run every second.
So far request _**(1)**_ has still not been successful, therefore, _**(2, 3 and 4)**_ as well.

When request _**(1)**_ is successful, the message "**_Message deleted_**" is inserted into the message session.

This is where the other requests respond, and all of them come back with the message array that contains the message "**_Message deleted_**".

The short video attached illustrates these explanations.
![Message deleted](https://github.com/user-attachments/assets/dd7ed79f-296d-4850-9d68-d34d18f0c565)

Reply:
> Here's how I figured out the bug:
> 
> Sending the request to delete the message: **_ajax_imap_delete_message_**(1) The time the request is being processed; The application sends other requests including: - **_ajax_imap_unsnooze_**(2) - **_ajax_imap_folder_display_**(3) - **_ajax_imap_unread_**(4)
> 
> Queries **_(2),(3), (4)_** run every second. So far request _**(1)**_ has still not been successful, therefore, _**(2, 3 and 4)**_ as well.
> 
> When request _**(1)**_ is successful, the message "**_Message deleted_**" is inserted into the message session.
> 
> This is where the other requests respond, and all of them come back with the message array that contains the message "**_Message deleted_**".
> 
> The short video attached illustrates these explanations. ![Message deleted](https://private-user-images.githubusercontent.com/62720246/359032876-dd7ed79f-296d-4850-9d68-d34d18f0c565.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ2NzQxNDcsIm5iZiI6MTcyNDY3Mzg0NywicGF0aCI6Ii82MjcyMDI0Ni8zNTkwMzI4NzYtZGQ3ZWQ3OWYtMjk2ZC00ODUwLTlkNjgtZDM0ZDE4ZjBjNTY1LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODI2VDEyMDQwN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM2NjZhZDQzZWYxMGVlMDFjMGZmOTI5ZDRlN2E3ZWI4ZGRiZDllNzEzZjY3MmZjYWU0ZWE1N2VmMzQ5MDhmNzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.n4SMg82TIm4Qgk-_R1Uksi3gVTEM4E1KiE6w3dqnw-o) [ ![Message deleted](https://private-user-images.githubusercontent.com/62720246/359032876-dd7ed79f-296d-4850-9d68-d34d18f0c565.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ2NzQxNDcsIm5iZiI6MTcyNDY3Mzg0NywicGF0aCI6Ii82MjcyMDI0Ni8zNTkwMzI4NzYtZGQ3ZWQ3OWYtMjk2ZC00ODUwLTlkNjgtZDM0ZDE4ZjBjNTY1LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODI2VDEyMDQwN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM2NjZhZDQzZWYxMGVlMDFjMGZmOTI5ZDRlN2E3ZWI4ZGRiZDllNzEzZjY3MmZjYWU0ZWE1N2VmMzQ5MDhmNzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.n4SMg82TIm4Qgk-_R1Uksi3gVTEM4E1KiE6w3dqnw-o) ](https://private-user-images.githubusercontent.com/62720246/359032876-dd7ed79f-296d-4850-9d68-d34d18f0c565.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ2NzQxNDcsIm5iZiI6MTcyNDY3Mzg0NywicGF0aCI6Ii82MjcyMDI0Ni8zNTkwMzI4NzYtZGQ3ZWQ3OWYtMjk2ZC00ODUwLTlkNjgtZDM0ZDE4ZjBjNTY1LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODI2VDEyMDQwN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM2NjZhZDQzZWYxMGVlMDFjMGZmOTI5ZDRlN2E3ZWI4ZGRiZDllNzEzZjY3MmZjYWU0ZWE1N2VmMzQ5MDhmNzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.n4SMg82TIm4Qgk-_R1Uksi3gVTEM4E1KiE6w3dqnw-o) [ ](https://private-user-images.githubusercontent.com/62720246/359032876-dd7ed79f-296d-4850-9d68-d34d18f0c565.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ2NzQxNDcsIm5iZiI6MTcyNDY3Mzg0NywicGF0aCI6Ii82MjcyMDI0Ni8zNTkwMzI4NzYtZGQ3ZWQ3OWYtMjk2ZC00ODUwLTlkNjgtZDM0ZDE4ZjBjNTY1LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODI2VDEyMDQwN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM2NjZhZDQzZWYxMGVlMDFjMGZmOTI5ZDRlN2E3ZWI4ZGRiZDllNzEzZjY3MmZjYWU0ZWE1N2VmMzQ5MDhmNzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.n4SMg82TIm4Qgk-_R1Uksi3gVTEM4E1KiE6w3dqnw-o)

Hello @kroky , did you manage to reproduce this?

Reply:
Then, I think the problem is $this->save_hm_msgs(); at the end of imap_delete_message handler module - we shouldn't store these in the session for subsequent requests to show since we are no reloading the page after deletion and ajax Hm_Ajax_request.done method successfully displays the router messages without reload.

Reply:
I think you misunderstood me. We shouldn't remove the save_hm_msgs functionality altogether - it is necessary when we have a redirect - otherwise, the message will be lost. I meant we should remove the call in places where we don't have a redirect - e.g. after message deletion.

Reply:
@christer77 can you please ping me when you do changes here as this PR is pretty old and we want to move forward with it?

Reply:
> I think you misunderstood me. We shouldn't remove the save_hm_msgs functionality altogether - it is necessary when we have a redirect - otherwise, the message will be lost. I meant we should remove the call in places where we don't have a redirect - e.g. after message deletion.

well noted

Reply:
But if we delete it as you say, how are we going to know that the action has succeeded, for example in the case of deleting the message as you said.

Reply:
Hm_Ajax_request.done method checks for router messages coming from ajax requests and shows them. So, unless there is a redirect, we should be already fine displaying those messages.